examples/nextjs-app-dir-validate-email/app/layout.tsx

```diff
@@ -9,11 +10,14 @@ export const metadata: Metadata = {
   description: "Generated by create next app",
 };
 
-export default function RootLayout({
+export default async function RootLayout({
   children,
 }: {
   children: React.ReactNode;
 }) {
+  // Opt-out of static generation for every page so the CSP nonce can be applied
```

Comment:
Isn't this quite a big change in how Sveltekit works? Could this happen at build time?

Reply:
This is actually a Next.js example.

The problem with Next.js is that it doesn't hash files that it produces statically (SvelteKit does this!). The result is that if you don't opt-out of static generation, then you need to set  `'self' 'unsafe-inline' 'unsafe-eval'` on the CSP as per their docs at https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy#without-nonces

This defeats the much of the purpose of the CSP, since untrusted JavaScript can run on the page. Here's the warnings and errors from https://csp-evaluator.withgoogle.com/:
<img width="978" alt="Screenshot 2024-11-18 at 8 26 44 AM" src="https://github.com/user-attachments/assets/1813ecf8-4cf6-4bc9-8f1d-bc4cb446a942">

I tried getting access to metadata about static/dynamic pages but it's not available at runtime. I believe this makes it impossible for us to figure out if the unsafe headers or nonce should be used and I opted for the secure nonce-based headers since we're trying to push safe defaults.

Reply:
Makes sense. We can document this reasoning and explain the tradeoffs.